File: Data Structures & Algorithms/binary-search/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        l=0
        mid = int(len(nums)/2)
        h = len(nums) - 1

        while l <= h:
            if nums[mid] < target:
                l=mid+1
                mid = int(l + (h- l)/2)
            elif nums[mid] > target:
                h=mid-1
                mid = int((l+h)/2)
            else:
                return mid

        return -1

File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
public:
    vector<vector<int>> res;

    void bk(int i, int currSum, vector<int>& nums,vector<int>& path,int target) {
        if (currSum == target) {
            res.push_back(path);
            return;
        }

        if (currSum > target || i >= nums.size()) {
            return;
        }

        //choice one: add current number
        path.push_back(nums[i]);
        bk(i,currSum + nums[i], nums,path,target);
        path.pop_back();

        //choice two: move on to next number
        bk(i + 1, currSum, nums, path, target);

    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> p = {};

        bk(0,0,nums,p,target);
        return res;
    }
};
